src/main/java/io/odpf/firehose/filter/JsonFilter.java

```diff
@@ -0,0 +1,168 @@
+import com.networknt.schema.ValidationMessage;
+import io.odpf.firehose.config.FilterConfig;
+import io.odpf.firehose.config.enums.FilterDataSourceType;
+import io.odpf.firehose.config.enums.FilterMessageFormatType;
+import io.odpf.firehose.consumer.Message;
+import io.odpf.firehose.metrics.Instrumentation;
+import org.apache.commons.lang3.reflect.MethodUtils;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import static io.odpf.firehose.config.enums.FilterDataSourceType.KEY;
+import static io.odpf.firehose.config.enums.FilterDataSourceType.NONE;
+import static io.odpf.firehose.config.enums.FilterMessageFormatType.PROTOBUF;
+
+/**
+ * JSON-based filter to filter protobuf/JSON messages based on rules
+ * defined in a JSON Schema string.
+ */
+public class JsonFilter implements Filter {
+
+    private final ObjectMapper objectMapper;
+    private final FilterDataSourceType filterDataSourceType;
+    private final FilterMessageFormatType messageType;
+    private Method protoParser;
+    private final Instrumentation instrumentation;
+    private JsonSchema schema;
+    private final JsonFormat.Printer jsonPrinter;
+
+    /**
+     * Instantiates a new Json filter.
+     *
+     * @param filterConfig    the consumer config
+     * @param instrumentation the instrumentation
+     */
+    public JsonFilter(FilterConfig filterConfig, Instrumentation instrumentation) {
+        objectMapper = new ObjectMapper();
+        JsonSchemaFactory schemaFactory = JsonSchemaFactory.getInstance(SpecVersion.VersionFlag.V7);
+        jsonPrinter = JsonFormat.printer().preservingProtoFieldNames();
+        messageType = filterConfig.getFilterMessageFormat();
+        filterDataSourceType = filterConfig.getFilterJsonDataSource();
+        String protoSchemaClass = filterConfig.getFilterJsonSchemaProtoClass();
+        String filterJsonSchema = filterConfig.getFilterJsonSchema();
+        this.instrumentation = instrumentation;
+        if (filterDataSourceType != NONE) {
+            try {
+                schema = schemaFactory.getSchema(filterJsonSchema);
+            } catch (Exception e) {
+                throw new FilterException("Failed to parse JSON Schema " + e.getMessage());
+            }
+        }
+        if (messageType == PROTOBUF) {
+            try {
+                protoParser = MethodUtils.getAccessibleMethod(Class.forName(protoSchemaClass), "parseFrom", byte[].class);
+            } catch (ClassNotFoundException e) {
+                throw new FilterException("Failed to load Proto schema class " + e.getMessage());
+            }
+        }
+    }
+
+    /**
+     * Log configs.
+     *
+     * @param filterConfig    the filter config
+     * @param instrumentation the instrumentation
+     */
+    public static void logConfigs(FilterConfig filterConfig, Instrumentation instrumentation) {
```

Comment:
we don't need it here. It should be in a Util class.

Reply:
fixed